ibis/sql/postgres/udf/api.py

```diff
@@ -0,0 +1,309 @@
+    -------
+    result : type
+        A new PostgresUDFNode subclass
+    """
+    definition = next(_udf_name_cache[name])
+    external_name = '{}_{:d}'.format(name, definition)
+    return type(external_name, (PostgresUDFNode,), fields)
+
+
+def existing_udf(name,
+                 input_types,
+                 output_type,
+                 schema=None,
+                 parameters=None):
+    """Create an ibis function that refers to an existing Postgres UDF already
+    defined in database
+
+    Parameters
+    ----------
+    name: str
+    input_types : List[DataType]
+    output_type : DataType
+    schema: str - optionally specify the schema that the UDF is defined in
+    parameters: List[str] - give names to the arguments of the UDF
+
+    Returns
+    -------
+    wrapper : Callable
+        The wrapped function
+    """
+    if parameters is None:
+        parameters = ['v{}'.format(i) for i in range(len(input_types))]
+    elif len(input_types) != len(parameters):
+        raise ValueError(
+            (
+                "Length mismatch in arguments to existing_udf: "
+                "len(input_types)={}, len(parameters)={}"
+            ).format(len(input_types), len(parameters))
+        )
+
+    udf_node_fields = collections.OrderedDict([
+        (name, Arg(rlz.value(type_)))
+        for name, type_ in zip(parameters, input_types)
+    ] + [
+        (
+            'output_type',
+            lambda self, output_type=output_type: rlz.shape_like(
+                self.args, dtype=output_type
+            )
+        )
+    ])
+    udf_node_fields['resolve_name'] = lambda self: name
+
+    udf_node = create_udf_node(name, udf_node_fields)
+
+    def _translate_udf(t, expr):
+        func_obj = sa.func
+        if schema is not None:
+            func_obj = getattr(func_obj, schema)
+        func_obj = getattr(func_obj, name)
+
+        sa_args = [t.translate(arg) for arg in expr.op().args]
+
+        return func_obj(*sa_args)
+
+    add_operation(udf_node, _translate_udf)
+
+    def wrapped(*args, **kwargs):
+        node = udf_node(*args, **kwargs)
+        return node.to_expr()
+
+    return wrapped
+
+
+class LineNums(ast.NodeVisitor):
+    """NodeVisitor for abstract syntax tree that notes the line numbers
+    of all decorator lines and (separately) all other node types"""
+    def __init__(self):
+        self.non_decorator_lines = list()
+        self.decorator_lines = list()
+
+    def visit_FunctionDef(self, node):
+        self.decorator_lines.extend(
+            [n.lineno for n in node.decorator_list]
+        )
+        for field1, node1 in ast.iter_fields(node):
+            if field1 != 'decorator_list' and isinstance(node1, ast.AST):
+                self.generic_visit(node1)
+
+    def generic_visit(self, node):
+        if hasattr(node, 'lineno'):
+            self.non_decorator_lines.append(node.lineno)
+        ast.NodeVisitor.generic_visit(self, node)
+
+
+def remove_decorators(funcdef_source):
```

Comment:
It's a bit concerning that we have to effectively delete code to make this work. The decorator API is nice, but I think having to use the `ast` module to manipulate the source is too much magic.

Reply:
We could leave the implementation as is, remove the decorator deletion code and document that use of the `udf` method as decorator will not work because of `$REASONS`.

Reply:
I can do that (remove decorator deletion code and add some documentation for `$REASONS`). I could also add a check that if it has decorators to throw an exception or warning.

Reply:
I see two patterns possible here. We could choose not to support decorator syntax but still have a two-step process to conform with the other UDF APIs:

```python
my_udf = ibis_client.udf([dt.int32], dt.int32)(my_function)
```

Or include the function object in the first call:

```python
my_udf = ibis_client.udf(my_function, [dt.int32], dt.int32)
```

I prefer the latter streamlined version. @cpcloud, would you be okay with that?

Reply:
I just pushed a new version that takes out the deletion and implements the latter non-decorator form (`ibis_client.udf(func, [dt.int32], dt.int32)`)